hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOp.java

```diff
@@ -134,9 +141,15 @@ public void afterRun() throws Exception {
         }
 
         final boolean shouldExecutePostJoinOp = preparePostOp(postJoinOp);
-        if (deferPartitionProcessing && getInternalHotRestartService() != null && getInternalHotRestartService().isEnabled()) {
-            getInternalHotRestartService().deferPostJoinOps(postJoinOp);
-            return;
+        if (getInternalHotRestartService() != null && getInternalHotRestartService().isEnabled()) {
+            // If recovery is still in progress do not execute now the post-join ops.
+            // For all operations, this is enforced in OperationRunnerImpl#checkNodeState. However,
+            // post-join ops are executed without any node state checks by OnJoinOp, therefore we
+            // need to explicitly defer execution after recovery from persistence is done.
+            if (deferPartitionProcessing || !getInternalHotRestartService().isStartCompleted()) {
```

Comment:
I think here should be `&&` not `||`?

Reply:
The `||` signifies that we need to defer post-join ops either because the `deferPartitionProcessing` flag is set or because recovery is not done yet. Comment updated in 4f840b6f9e7a8b6f775e7ee9951358689ce5d022 for clarification.